### ISSUE TITLE
High rate of generating repeat fingerprint id

### ISSUE THREAD

**Issue by weixing02** [fingerprintjs/fingerprintjs#665]
### Scenario

- What's the scenario, what happens and what did you expect to happen?
I am using fingerprintjs to produce unique id for our customers, and repeated id cannot get award from our service. But i checked log and found that 80,000 / 200,000 records are repeated! 

Here is the data generated from fingerprintjs. The device type, ip, vaid, imei is different, but they all have the same fingerprint id, which is wiered and caused a bad result... 

![image](https://user-images.githubusercontent.com/7108005/120481781-0a232500-c3e3-11eb-8438-c40b840243b1.png)

Is that my wrong usage caused this problem? Please help me to figure this out. I'll appreciate.

I followed the documentation and used it in my page initilization.
``` 
const fpPromise = FingerprintJS.load()

      const fp = await fpPromise
      const result = await fp.get()
```

- What device and browser are you using?
Our customers can be using different devices and browsers, as you can see in the picture. But the same fingerprintid are generated. 

- What version of fingerprintjs are you using?
 "@fingerprintjs/fingerprintjs": "^3.1.2"

**Comment by weixing02** [comment]
But there seems have a wiered pattern, that all the records are all android 10 + Chrome/78.0.3904.96 or android 11 + Chrome/83.0.4103.106.

**Comment by makma** [comment]
Hi 👋 @weixing02, thank you for your question!
- Could you please try to narrow down all the correlated /common traits/properties? Do I understand correctly that only users with Android 10 + Chrome/78.0.3904.96 or android 11 + Chrome/83.0.4103.106 got the same false-positive fingerprint? Does it work for the other visitors just fine?
- Do I understand correctly that 80k fingerprints are the same? -  does this imply the rest  - 120k are different fingerprints?
Cannot be your data poisoned? I mean, are you sure those 80k fingerprints come really from 80k users (maybe a glitch in your reporting logic, bots, anything else)? Can you distinguish amongst them (e.g. with login)?
- Your provided initialization code snippet looks good to me, moreover, it probably works as intended considering the fact you are satisfied with the rest of 120k fingerprints.

**Most importantly, could you provide us with steps to reproduce with some minimal setup? This would help a lot.**🎉

**Comment by Finesse** [comment]
We don't use user-agent because it's often fake. JavaScript has no access to IMEI so we can't use it for identification. It's ok that many devices get the same browser fingerprint, especially when they are factory reset devices or a bot farm. Also, when you know how the fingerprinting works (you can, because the source code is open), you can make a custom browser that produces the same fingerprint on purpose.

IP address is not a part browser fingerprinting. You can combine fingerprint with IP address and other factors to achieve better accuracy. This is what FingerprintJS Pro does.

**Comment by weixing02** [comment]
> Hi 👋 @weixing02, thank you for your question!
> 
> * Could you please try to narrow down all the correlated /common traits/properties? Do I understand correctly that only users with Android 10 + Chrome/78.0.3904.96 or android 11 + Chrome/83.0.4103.106 got the same false-positive fingerprint? Does it work for the other visitors just fine?
> * Do I understand correctly that 80k fingerprints are the same? -  does this imply the rest  - 120k are different fingerprints?
>   Cannot be your data poisoned? I mean, are you sure those 80k fingerprints come really from 80k users (maybe a glitch in your reporting logic, bots, anything else)? Can you distinguish amongst them (e.g. with login)?
> * Your provided initialization code snippet looks good to me, moreover, it probably works as intended considering the fact you are satisfied with the rest of 120k fingerprints.
> 
> **Most importantly, could you provide us with steps to reproduce with some minimal setup? This would help a lot.🎉**

Hi, thanks for the reply! 
- Sorry that we cannot narrow down the correlated traits/properties. We stopped using fingerprintjs on our website and properties in the table are all we got. I am not sure if all the repeated ids came from Android 10 + Chrome/78.0.3904.96 or android 11 + Chrome/83.0.4103.106. 
- Sorry for not being clear, 80k ids are repeated, but i calculated the counts of each repeated id, here is the distribution:
8652::[40fb5b3523a20b963196a5ead0be5efa]
6117::[33c383ac51689a7087f286270475b8e1]
5045::[a37193af0db77013a3fcf2996f549e95]
4767::[c0f8adc2ae9ae349174a1e6399f0b019]
4723::[40b4778585eccc67eead09d53b3be94a]
3639::[10e2078dab7f29cbd0b80cad29892062]
3507::[fef0f7979a26c787515dd36cae1fcccb]
3139::[f05f1022d33b5114cd02355cf74fa084]
2905::[f80c960898d850a566a662d1142c28ae]
2818::[2b01e0c8f2bb480ae1cf5d3f8cba8560]
2657::[4e96c169ac20a155ae1a7eef0d9aa571]
2235::[e45a034a5ffa9894cfc6a9ba34174738]
1492::[5f9d226977e20c9e6a45a817268fb4a0]
1408::[1c8ff7a7760c47774d3550f4d0ee3ca0]
1287::[aa9206f1f3a1eb920554fd131c17f4ec]
1035::[76fa692cb8ca99602ff3c49ed6aa284a]
1008::[db163287848fa296386ddb69f8b93f87]
810::[451c35db44ce2f096c1afbb59d1b182a]
732::[2cab9a925951d9f073fffd9c9075ff8d]
679::[c23ba03b2a56e544d351fa237b61a179]
523::[44801094074fce26ae42b328b7e27d8a]
509::[67f5a5fdb978c087c6f9274fd0b9640f]
502::[e749488368ad481749fb7dbdda5a497b]
466::[d21d119490a27966515bdb4c452cec9d]
457::[f3cd1091a5ebb02ebb6937c7fd8f72c0]
other 120k are different unique ids.
as you can see in the table, they all have diffrent ip, vaid, device type, so i am sure that these ids came from different users.

- Still not find the regular steps to reproduce, but we do reproduced it on several android device webview.

**Comment by weixing02** [comment]
> We don't use user-agent because it's often fake. JavaScript has no access to IMEI so we can't use it for identification. It's ok that many devices get the same browser fingerprint, especially when they are factory reset devices or a bot farm. Also, when you know how the fingerprinting works (you can, because the source code is open), you can make a custom browser that produces the same fingerprint on purpose.
> 
> IP address is not a part browser fingerprinting. You can combine fingerprint with IP address and other factors to achieve better accuracy. This is what FingerprintJS Pro does.

Thank you for the comment! This could be useful and i found the similar issue in #612

**Comment by makma** [comment]
Hi @weixing02, Thank you for the provided data. I reckon Finesse is right. It seems the fingerprinting works as intended for you and given hashes just represent identical devices and browsers combinations from the FingerprintJS logic point of view - I'm afraid this is a by-design behavior of browser fingerprint. If you require better accuracy you can try some of the techniques mentioned in #612.
Additionally, your note about WebViews is very valuable - I believe this comment might be beneficial for you - https://github.com/fingerprintjs/fingerprintjs/issues/612#issuecomment-777259567.
Long story short - if you want to get better accuracy you should combine fingerprint with another technique to get higher reliability or take a look at the [PRO](https://github.com/fingerprintjs/fingerprintjs#switch-to-pro-version-for-free-to-get-995-identification-accuracy) version.

**Comment by makma** [comment]
I’m closing this issue now because of inactivity.

@weixing02 we can always reopen it though! :) Please update the issue if needed.
